### DISCUSSION TITLE
STIG MANAGER on SIPRNET?

### DISCUSSION THREAD

**Discussion by st3v1s** [NUWCDIVNPT/stig-manager#1000] [Q&A]
I'm not having any luck getting the api container to start on a server on SIPRNET.  I exported/imported the auth, mysql, and api images from our  NIPR STIG MANAGER and our docker-compose.yml looks similar to the one provided.

Is there a SIPRnet version such that after mysql and keycloak are up, it can reach out to DoD Cyber Exchange on SIPRnet to import stigs?

The Keycloak and Mysql containers startup with no errors but the stig-manager api either throws an error if I have init: true in the docker-compose.yml file or it errors out trying to establish a connection.

Any help would be greatly appreciated!

**Comment by cd-rite** [comment]
Hi @st3v1s Ah, that explains it.... STIGMan won't be able to reach out over the internet to import STIGs from DISA on the SIPR side.
You'll have to set that to False, and import them manually once it's up.

**Reply by st3v1s** [reply]
Yes understand that but still get error when trying to start up docker-api-1.  I see that keycloak and mysql are up and ready. But api tries over and over and eventually shutsdown with error setting up dependencies.

Errors I see:  mysql, preflight, data, success: false, message: connect EHOSTUNREACH IP:50001
                     component, oidc, discovery, data, success: false, message: connect EHOSTUNREACH: IP/realms/stigman/.well-known/openid-configuration. , message: connect EHOSTUNREACH: IP:8080

**Comment by cd-rite** [comment]
Hi @st3v1s  If you are using an orchestration like docker-compose, you probably wouldn't have IPs in the addresses for the API and Keycloak.... They are usually aliased to something defined in the file, as in our demo orchestration: https://github.com/NUWCDIVNPT/stigman-orchestration/blob/main/docker-compose.yml

The api config section then will reference the aliases, instead of IPs:

      - STIGMAN_OIDC_PROVIDER=http://keycloak:8080/realms/stigman
      - STIGMAN_DB_HOST=mysql

I'd compare what you are running to our demo to see if they relate.
Beyond that, you may need to ensure you have provided the proper certs for that environment, though I think that would likely be a different error.

**Comment by jeremyatourville** [comment]
@st3v1s Did you have any luck getting your issue worked out?  I am working out a similar issue (trying to make the app work in SIPR)

**Reply by st3v1s** [reply]
In the end what worked for me was cloning our working version on NIPR and then moving the VM to SIPR. Just had to change hostname and network settings.

**Reply by jeremyatourville** [reply]
Glad to hear you got it working!  Hope it works out they way you wanted it to.

**Reply by OlympianRange** [reply]
@st3v1s how'd you get it working on nipr? I've been spinning my wheels on this for weeks. I was initially able to get the demo build working. But now I can't get a actual working build going. I'm in the same boat, want this done to be able to move it to an airgapped network.

**Reply by jeremyatourville** [reply]
@OlympianRange Try taking a look at https://github.com/NUWCDIVNPT/stigman-orchestration/pulls  That should get you going in the right direction.
I recently added that as a PR because other users are facing the same challenge. 

Use the stigman-orchestration repo as a starting point and then make the changes listed in the PR for nginx.conf and docker-compose.yml files

**Reply by OlympianRange** [reply]
@jeremyatourville cool thanks! I'll give it try today. Did you run this on a windows host or linux?

**Reply by jeremyatourville** [reply]
Linux, but Windows should work as well.

**Reply by cd-rite** [reply]
Hi @jeremyatourville Thanks for those PRs... Rather than incorporate them into our repo, however, we'd like to update the Docs to point to your repo as one possible deployment approach.  As more users find and share solutions that work for them, it should be helpful to accumulate multiple examples for other users to refer to!
Thanks for helping out @OlympianRange !

**Reply by jeremyatourville** [reply]
@cd-rite I don't actually have a repo.  The point of creating the PR was to provide documentation that would allow users to have a shared deployment.  I think that would be a common use case.  Only you as the developers would actually know if that use case scenario is requested often. Feel free to use however you see fit.  I was simply trying to fill a gap that I saw.  No problem with assisting OlympianRange, glad to help if I can.

**Reply by OlympianRange** [reply]
@jeremyatourville @cd-rite thanks for the help, i appreciate it! I'm still having trouble with this current method of deployment, I can see how it "should" work, but i keep getting an error back on the permissions for /tmp directory with the certs. This is likely a user error on my part though.

I figure it's worth bringing up, does anyone have a down and dirty implementation? Just allow folks to be able to log into stigman with username and pass?

**Reply by jeremyatourville** [reply]
@OlympianRange yes, you need to get the certs to work.  If you have having permission issues that will definitely need to be fixed.  I am confident when you have fixed the cert issues the application will run as expected.  You describe the exact use case scenario that I already solved- getting stigman to work with username and password.  Keep working on it, you are very close!  :-)  I'll review my implementation to see what I did with permissions on the certs.

**Reply by jeremyatourville** [reply]
@OlympianRange please provide the output of `docker-compose logs -f` when you start up the app.  That will help us see exactly what is going wrong.  Also, it will be very helpful if you can post a copy of your docker-compose.yml file as you have it right now.

**Reply by OlympianRange** [reply]
[docker-compose.txt](https://github.com/NUWCDIVNPT/stig-manager/files/12175173/docker-compose.txt)

@jeremyatourville i attached my docker-compose.yml as .txt (just to be clear, i'm new to development and using github and i'm not sure if i attached it right), the error I get when trying to run it is "yaml: line 2: did not find expected key" at first i figured i might've had a space or tab somewhere i shouldn't have. I went through tidying up what I could and still get the same error.

Now as far as the issue with the /tmp/truststore.p12, I forgot one of the things I had tried yesterday was running the containers individually,  I didn't get passed getting the keycloak container to run, my attempt was running this command:

```
docker run -p 8080:8080 -p 8443:8443 
\-e KEYCLOAK_ADMIN=admin 
\-e KEYCLOAK_ADMIN_PASSWORD=Pa55w0rd 
\-e KC_PROXY=edge 
\-e KC_HOSTNAME=my-hostname 
\-e KC_HOSTNAME_ADMIN_URL=https://localhost/kc/ 
\-e KC_SPI_TRUSTSTORE_FILE_FILE=/tmp/truststore.p12 
\-e KC_SPI_TRUSTSTORE_FILE_PASSWORD=password 
\-v ./certs/dod/Certificates_PKCS7_v5.9_DoD.pem.p12:/tmp/truststore.p12 
\-v ./kc/stigman_realm.json:/opt/keycloak/data/import/stigman_realm.json 
\-v ./kc/h2:/opt/keycloak/data/h2 quay.io/keycloak/keycloak:latest start --import-realm
``` 

Error Received from that command:

2023-07-26 18:03:11,997 INFO  [io.quarkus.deployment.QuarkusAugmentor] (main) Quarkus augmentation completed in 14448ms
ERROR: Unexpected error when starting the server in (development) mode
ERROR: Failed to start quarkus
ERROR: Failed to initialize TruststoreProviderFactory: /tmp/truststore.p12, truststore type: PKCS12
ERROR: /tmp/truststore.p12 (Is a directory)
For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.

**Reply by jeremyatourville** [reply]
@OlympianRange Oh!!! You aren't really using docker-compose.  You are using docker run (which is very different).  Take a look at my repo - https://github.com/jeremyatourville/stigman-orchestration/blob/main/docker-compose.yml  Do you have docker compose installed?  If you do, you can copy the docker-compose.yml file into your directory.  Then you can run `docker -compose up -d && docker-compose logs -f` and all the containers will attempt to start.  It will also output logs so we can see what is going wrong.

**Reply by OlympianRange** [reply]
Okay yeah, big difference from yesterday, when I tried the same yesterday it didn't work period. But now everything started running, however it looks like I'm getting errors from Keycloak and permissions to it's data volume. And possibly and potentially an issue with the nginx.conf file. I feel like you might have mentioned that somewhere in your README so I'm going to go back and take a look.

Also I'm assuming in the docker-compose.yml file I would change the "Your_DNS_Hostname" to the name of the host i'm running the containers from correct? I tried both ways (with and with editting) and got the same error. I'm wondering if I can just throw the IP of the VM I'm working off of on there.

**Reply by jeremyatourville** [reply]
For the data volume I presume you mean this one:
- ./kc/h2:/opt/keycloak/data/h2
If so, you can set permissions
chmod 764 /kc/h2
chown -R 1000 h2

> Also I'm assuming in the docker-compose.yml file I would change the "Your_DNS_Hostname" to the name of the host i'm running the containers from correct?
Correct but there is a relationship between nginx , docker-compose and having certs in the correct locations.  Be warned that it may not work as expected.  It might appear to work and still fail in practical use.  If you don't care about SSL/TLS for nginx at all I'd comment out any lines related to certs.  That goes for both nginx.conf and docker-compose.yml.  (They are related to each other.  If you change one, you need to change the other in the appropriate locations.)

**Comment by cd-rite** [comment]
@jeremyatourville You do! Your fork is a repo that is a copy of our repo: https://github.com/jeremyatourville/stigman-orchestration
You could merge your proposed changes into your own "main" branch, get it the way you want it, and then we will add something like a "Community Solutions" section to our docs with a link to your fork.  Or you could have a specific branch with your approach. Just let me know!
Thanks!

**Reply by jeremyatourville** [reply]
@cd-rite   I have very little development background so I didn't know that. You explained it well.  I will try to attempt what you suggested.  I may need a little assistance if I run into issues but I'll let you know and give it a shot.  Thanks!

**Reply by jeremyatourville** [reply]
@cd-rite   OK, I think I have it all set.  My main branch reflects the changes from the PR and provides some guidance for a use case scenario with username/password, sharing the application among multiple users and using DNS to get to the app.  See my updated fork - [https://github.com/jeremyatourville/stigman-orchestration/tree/main](https://github.com/jeremyatourville/stigman-orchestration/tree/main)

**Reply by cd-rite** [reply]
Great! Thanks @jeremyatourville 
I've added added a "Community Solutions" section to Readme with link to your repo, and closed your PRs.
Thanks for the contribution!
https://github.com/NUWCDIVNPT/stigman-orchestration#community-solutions

I haven't had a chance to test it, but your Readme does still mention CAC.... Are you authenticating CAC at the reverse proxy, but then allowing username/pw when you hit keycloak?

**Reply by OlympianRange** [reply]
@cd-rite do you know what environment variables were used for the keycloak image that was made for demo purposes? keycloak seems to be the only piece of this whole thing i have yet to get functioning.

**Reply by jeremyatourville** [reply]
@cd-rite   I am authenticating with username/password at keycloak, no auth at the reverse proxy is required.  I overlooked that section of the Readme but I have fixed that part of the documentation now.  There should not be any more references to CAC.  Let me know if you spot any other obvious errors.

**Reply by jeremyatourville** [reply]
@OlympianRange Where are you still having trouble with keycloak?  Run this command and paste the outuput here, it will help us determine what is going wrong.  `docker-compose down && docker-compose up -d && docker-compose logs -f`  Have you done any configuration in the keycloak gui to set up your users?

**Reply by OlympianRange** [reply]
@jeremyatourville 
[v3output.txt](https://github.com/NUWCDIVNPT/stig-manager/files/12253099/v3output.txt)
[docker-compose .txt](https://github.com/NUWCDIVNPT/stig-manager/files/12253105/docker-compose.txt)


Attached is the output and the docker compose file. Before this attempt, I tried running everything on baremetal, it didn't entirely work, since regardless of the commands I used, I couldn't get keycloak to persist it's configs, and it would not let me start keycloak in prod mode.

**Reply by cd-rite** [reply]
kinda looks like keycloak can't write to its volume.... maybe check the permissions on the host system folder?:

keycloak_1  | Caused by: java.io.FileNotFoundException: /opt/keycloak/data/h2/keycloakdb.trace.db (Permission denied)
keycloak_1  | 	at java.base/java.io.FileOutputStream.open0(Native Method)
keycloak_1  | 	at java.base/java.io.FileOutputStream.open(FileOutputStream.java:298)

keycloak_1  | 2023-08-03 16:20:10,031 WARN  [io.agroal.pool] (agroal-11) Datasource '<default>': Error opening database: "Could not save properties /opt/keycloak/data/h2/keycloakdb.lock.db" [8000-197]

**Reply by jeremyatourville** [reply]
As a quick test you can also comment out that line in the docker-compose file and see if the container starts and stays running.  If it does, then you know you need to fix the permissions.

  ```
  volumes:
      - ./certs/dod/certificates_pkcs7_v5_12_dod_pem.p12:/tmp/truststore.p12
      - ./kc/stigman_realm.json:/opt/keycloak/data/import/stigman_realm.json
#      - ./kc/h2:/opt/keycloak/data/h2
```

**Comment by cd-rite** [comment]
@OlympianRange  Our demo described here just uses the vanilla Keycloak container from RedHat (v19.0.2), nothing we built specifically:
https://github.com/NUWCDIVNPT/stigman-orchestration


If you are talking about the demo shown here:
https://hub.docker.com/r/nuwcdivnpt/stig-manager
that does use a very very very simple config of keycloak with built in accounts that you wouldn't want to use for anything other than looking at the app locally. 
But, that container is built from the Dockerfile here: https://github.com/NUWCDIVNPT/stig-manager-auth/blob/main/Dockerfile
and just shows that we preset the admin password and imported a realm from that repo.
